lib/schemes/VirtualUrlPlugin.js

```diff
@@ -0,0 +1,139 @@
+ */
+
+/**
+ * Normalizes a virtual module definition into a standard format
+ * @param {string | ((resourcePath: string, loaderContext: LoaderContext<EXPECTED_ANY>) => Promise<string> | string) | VirtualModule} virtualModule The virtual module to normalize
+ * @returns {VirtualModule} The normalized virtual module
+ */
+function normalizeModule(virtualModule) {
+	if (typeof virtualModule === "string") {
+		return {
+			type: "",
+			source() {
+				return virtualModule;
+			}
+		};
+	} else if (typeof virtualModule === "function") {
+		return {
+			type: "",
+			source: virtualModule
+		};
+	}
+	return virtualModule;
+}
+
+/**
+ * Normalizes all virtual modules with the given scheme
+ * @param {VirtualModules} virtualModules The virtual modules to normalize
+ * @param {string} scheme The URL scheme to use
+ * @returns {{[key: string]: VirtualModule}} The normalized virtual modules
+ */
+function normalizeModules(virtualModules, scheme) {
+	return Object.keys(virtualModules).reduce((pre, id) => {
+		pre[`${scheme}:${id}`] = normalizeModule(virtualModules[id]);
+		return pre;
+	}, /** @type {{[key: string]: VirtualModule}} */ ({}));
+}
+
+/**
+ * @typedef {object} VirtualUrlPluginOptions
+ * @property {VirtualModules} modules - The virtual modules
+ * @property {string=} scheme - The URL scheme to use
+ */
+
+class VirtualUrlPlugin {
+	/**
+	 * @param {VirtualModules} modules The virtual modules
+	 * @param {string=} scheme The URL scheme to use
+	 */
+	constructor(modules, scheme) {
+		this.scheme = scheme || DEFAULT_SCHEME;
+		this.modules = normalizeModules(modules, this.scheme);
+	}
+
+	/**
+	 * Apply the plugin
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		const scheme = this.scheme;
+
+		compiler.hooks.compilation.tap(
+			PLUGIN_NAME,
+			(compilation, { normalModuleFactory }) => {
+				normalModuleFactory.hooks.resolveForScheme
+					.for(scheme)
+					.tap(PLUGIN_NAME, resourceData => {
+						const url = new URL(resourceData.resource);
+						const path = url.pathname;
+
+						const virtualModule = this.findVirtualModuleById(
+							resourceData.resource
+						);
+						const type = virtualModule.type;
+						resourceData.path = path + type;
+						resourceData.resource = `${scheme}:${path}`;
+						return true;
+					});
+
+				const hooks = NormalModule.getCompilationHooks(compilation);
+				hooks.readResource
+					.for(scheme)
+					.tapAsync(PLUGIN_NAME, async (loaderContext, callback) => {
+						const { resourcePath } = loaderContext;
+						try {
+							const virtualModule = this.findVirtualModuleById(resourcePath);
+							const sourceContent = await virtualModule.source(
+								resourcePath,
+								loaderContext
+							);
+							callback(null, sourceContent);
+						} catch (err) {
+							callback(/** @type {Error} */ (err));
+						}
+					});
```

Comment:
What do you think to add generated code to `valueDependency` (and maybe to allow to set them using loaderContext, yeah, maybe we need API for it), so when valud was changed (it can be read from external file) we will invalidate our code?

Reply:
The version I’ve implemented now stores the user-provided version in valueDependencies, with the parameter type and inspiration drawn from DefinePlugin.RuntimeValue.version. When the version changes from the previous one, it triggers a build, which then calls the user-provided factory function source.
https://github.com/webpack/webpack/pull/19508/files#diff-6060d03f2a91f56b93302860b771391b8bc6511144fab0a7cd80e326c9fbb1dcR15